### PR TITLE
Implementation of `mockCall` cheat code

### PR DESCRIPTION
**Pull request by lisandrasilva**
Closes https://github.com/runtimeverification/kontrol/issues/285.

This PR implements the `mockCall` cheatcode with the following signature:
```solidity
function mockCall(address where, bytes calldata data, bytes calldata retdata) external;
```

### CODE REVIEW

File: src/kontrol/kdist/cheatcodes.md [hunk 1]
```diff
@@ -907,6 +913,53 @@ Otherwise, throw an error for any other call to the Foundry contract.
...
+           <mockAddress> ACCTTO </mockAddress>
+           <mockValues>...  CALLDATA |-> RETURNDATA ...</mockValues>
+         </mockCall>
+         requires #range(LM, ARGSTART, ARGWIDTH) ==K #range(CALLDATA,0,ARGWIDTH)
+      [priority(30)]
+```
+
+```k
+    syntax KItem ::= "#updateOutput" Bytes [klabel("foundry_updateRevertOutput")]
+ // -------------------------------------------------------------------------------------
```
Comment (anvacaru) [inline src/kontrol/kdist/cheatcodes.md:958]:
```suggestion
 // -----------------------------------------------------------------------------
```

File: src/kontrol/kdist/cheatcodes.md [hunk 2]
```diff
@@ -907,6 +913,53 @@ Otherwise, throw an error for any other call to the Foundry contract.
...
+    rule [foundry.set.mockCall]:
+         <k> #next [ _OP:CallOp ] ~> (. =>  #setLocalMem RETSTART RETWIDTH RETURNDATA) ~> #execute ... </k>
+         <localMem> LM </localMem>
+         <wordStack> _ : ACCTTO : _ : ARGSTART : ARGWIDTH : RETSTART : RETWIDTH : _WS </wordStack>
+         <mockCall>
+           <mockAddress> ACCTTO </mockAddress>
+           <mockValues>...  CALLDATA |-> RETURNDATA ...</mockValues>
+         </mockCall>
+         requires #range(LM, ARGSTART, ARGWIDTH) ==K #range(CALLDATA,0,ARGWIDTH)
+      [priority(30)]
```
Comment (anvacaru) [inline src/kontrol/kdist/cheatcodes.md:942]: I think that here you could have something like this instead. Basically saying that if you have a mocked value for a given address, skip executing the `CALL` opcode and just set the mocked value in the local memory and output cell.
```suggestion
    rule [foundry.set.mockCall]:
         <k> #next [ _OP:CallOp ] =>  #setLocalMem RETSTART RETWIDTH RETURNDATA  ... </k>
         <localMem> LM </localMem>
         <wordStack> _ : ACCTTO : _ : ARGSTART : ARGWIDTH : RETSTART : RETWIDTH : _WS </wordStack>
         <mockCall>
           <mockAddress> ACCTTO </mockAddress>
           <mockValues>...  CALLDATA |-> RETURNDATA ...</mockValues>
         </mockCall>
         requires #range(LM, ARGSTART, ARGWIDTH) ==K #range(CALLDATA,0,ARGWIDTH)
      [priority(30)]
```

File: src/kontrol/kdist/cheatcodes.md (hunk 1 above)
Comment (anvacaru) [reply src/kontrol/kdist/cheatcodes.md:958]: Also, where is this function used?

File: src/kontrol/kdist/cheatcodes.md [hunk 3]
```diff
@@ -1316,6 +1369,29 @@ If the production is matched when no prank is active, it will be ignored.
         </whitelist>
 ```
 
+- `#setMockCall MOCKADDRESS MOCKCALLDATA MOCKRETURN` will update the `<mockcalls>` mapping for the given account.
+
+```k
+    syntax KItem ::= "#setMockCall" Account Bytes Bytes [klabel(foundry_setMockCall)]
+ // ------------------------------------------------------------------------
```
Comment (anvacaru) [inline src/kontrol/kdist/cheatcodes.md:1376]: Formatting.
```suggestion
 // --------------------------------------------------------------------------------
```

File: src/tests/integration/test-data/foundry/test/MockCallTestFoundry.t.sol
Comment (anvacaru) [inline src/tests/integration/test-data/foundry/test/MockCallTestFoundry.t.sol:1]: We already have a few tests in place for this cheat code in [MockCallTest.t.sol](https://github.com/runtimeverification/kontrol/blob/master/src/tests/integration/test-data/foundry/test/MockCallTest.t.sol) are those passing? Do we need these two new files?

File: src/kontrol/kdist/cheatcodes.md (hunk 1 above)
Comment (lisandrasilva) [reply src/kontrol/kdist/cheatcodes.md:958]: It's not being used anymore. I will delete it.

File: src/kontrol/kdist/cheatcodes.md (hunk 2 above)
Comment (lisandrasilva) [reply src/kontrol/kdist/cheatcodes.md]: That was my first approach, but then I noticed (and tested) that actually if the function called changes the state of the contract, those changes are made, which means that the function is executed, but instead, it returns the value provided in the `mockCall`.

File: src/tests/integration/test-data/foundry/test/MockCallTestFoundry.t.sol
Comment (lisandrasilva) [reply src/tests/integration/test-data/foundry/test/MockCallTestFoundry.t.sol:1]: These tests were copied from the Foundry repo. I added them because they are more complete in the sense that they test scenarios that our tests were not catching.

File: src/kontrol/kdist/cheatcodes.md (hunk 2 above)
Comment (anvacaru) [reply src/kontrol/kdist/cheatcodes.md]: I see. Then you can leave them as they are. :+1:

File: src/tests/integration/test-data/foundry/test/MockCallTestFoundry.t.sol
Comment (anvacaru) [reply src/tests/integration/test-data/foundry/test/MockCallTestFoundry.t.sol:1]: Hmm, could you instead merge the two test files into a single one?

File: src/kontrol/kdist/cheatcodes.md [hunk 4]
```diff
@@ -907,6 +913,46 @@ Otherwise, throw an error for any other call to the Foundry contract.
...
+    //        => #loadAccount #asWord(#range(ARGS, 0, 32))
+    //        ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, 96, #asWord(#range(ARGS, 64, 32))) false ... </k>
+    //  requires SELECTOR ==Int selector ( "mockCall(address,uint256,bytes,bytes)" )
+```
+
+```k
+    rule [foundry.set.mockCall]:
+         <k> #next [ _OP:CallOp ] ~> (. =>  #setLocalMem RETSTART RETWIDTH RETURNDATA) ~> #execute ... </k>
+         <localMem> LM </localMem>
+         <wordStack> _ : ACCTTO : _ : ARGSTART : ARGWIDTH : RETSTART : RETWIDTH : _WS </wordStack>
```
Comment (anvacaru) [inline src/kontrol/kdist/cheatcodes.md:936]: To avoid warnings, you can replace unused args with `_`, or prefix their name with underscore.
```suggestion
         <wordStack> _ : ACCTTO : _ : ARGSTART : _ : RETSTART : RETWIDTH : _WS </wordStack>
```

File: src/kontrol/kdist/cheatcodes.md [hunk 5]
```diff
@@ -907,6 +913,46 @@ Otherwise, throw an error for any other call to the Foundry contract.
...
+           <mockAddress> ACCTTO </mockAddress>
+           <mockValues>...  CALLDATA |-> RETURNDATA ...</mockValues>
+         </mockCall>
+         requires #range(LM, ARGSTART, lengthBytes(CALLDATA)) ==K CALLDATA
+      [priority(30)]
+
+      rule [foundry.set.mockCall2]:
+         <k> #next [ _OP:CallSixOp ] ~> (. =>  #setLocalMem RETSTART RETWIDTH RETURNDATA) ~> #execute ... </k>
+         <localMem> LM </localMem>
+         <wordStack> _ : ACCTTO : ARGSTART : ARGWIDTH : RETSTART : RETWIDTH : _WS </wordStack>
```
Comment (anvacaru) [inline src/kontrol/kdist/cheatcodes.md:947]:
```suggestion
         <wordStack> _ : ACCTTO : ARGSTART : _ : RETSTART : RETWIDTH : _WS </wordStack>
```

File: src/tests/integration/test-data/foundry/test/MockCallTestFoundry.t.sol
Comment (lisandrasilva) [reply src/tests/integration/test-data/foundry/test/MockCallTestFoundry.t.sol:1]: Yes, I'll work on that.

Comment (lisandrasilva) [reply src/tests/integration/test-data/foundry/test/MockCallTestFoundry.t.sol:1]: Actually, this test from [MockCallTest.t.sol](https://github.com/runtimeverification/kontrol/blob/9ef828b5d29126a1f9bbdbb17ccf1cf8a0f657f5/src/tests/integration/test-data/foundry/test/MockCallTest.t.sol#L11) is failing. I think is related with the call being made to `address(0)`  but I need to investigate more.

File: src/kontrol/kdist/cheatcodes.md [hunk 6]
```diff
@@ -907,6 +913,48 @@ Otherwise, throw an error for any other call to the Foundry contract.
       [owise]
 ```
 
+
```
Comment (anvacaru) [inline src/kontrol/kdist/cheatcodes.md:916]: Please add a description of the cheat code, including the current limitations that it has.

File: src/kontrol/kdist/cheatcodes.md [hunk 7]
```diff
@@ -907,6 +913,48 @@ Otherwise, throw an error for any other call to the Foundry contract.
...
+          ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, #asWord(#range(ARGS, 32, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 32, 32)), 32))) #range(ARGS, #asWord(#range(ARGS, 64, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 64, 32)), 32)))
+         ...
+         </k> 
+      requires SELECTOR ==Int selector ( "mockCall(address,bytes,bytes)" )
+
+    //rule [foundry.call.mockCallWithMsgValue]:
+    //     <k> #call_foundry SELECTOR ARGS
+    //        => #loadAccount #asWord(#range(ARGS, 0, 32))
+    //        ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, 96, #asWord(#range(ARGS, 64, 32))) false ... </k>
+    //  requires SELECTOR ==Int selector ( "mockCall(address,uint256,bytes,bytes)" )
```
Comment (anvacaru) [inline src/kontrol/kdist/cheatcodes.md:931]: Please remove this commented code.
```suggestion
```

File: src/kontrol/kdist/cheatcodes.md [hunk 8]
```diff
@@ -1316,6 +1364,56 @@ If the production is matched when no prank is active, it will be ignored.
...
+                  <mockValues> .Map [ MOCKCALLDATA <- MOCKRETURN ] </mockValues>
+               </mockCall>
+           )
+           ...
+         </mockCalls>
+```
+
+```k
+    syntax KItem ::= "#clearMockCall" Int Int Bytes [klabel(foundry_clearMockCall)]
+ // -------------------------------------------------------------------------
```
Comment (anvacaru) [inline src/kontrol/kdist/cheatcodes.md:1410]: Formatting.
```suggestion
    syntax KItem ::= "#clearMockCall" Int Int Bytes [klabel(foundry_clearMockCall)]
 // -------------------------------------------------------------------------------
```

File: src/kontrol/kdist/cheatcodes.md [hunk 9]
```diff
@@ -1316,6 +1364,56 @@ If the production is matched when no prank is active, it will be ignored.
...
+    rule <k> #clearMockCall RETSTART RETWIDTH RETURNDATA => #setLocalMem RETSTART RETWIDTH RETURNDATA ... </k>
+         <statusCode> _ => EVMC_SUCCESS </statusCode>
+         <wordStack> _ : WS => 1 : WS </wordStack>
+```
+
+
```
Comment (anvacaru) [inline src/kontrol/kdist/cheatcodes.md:1416]: Formatting.
```suggestion
```

**Review by anvacaru — COMMENTED** [review]
Left a few comments. Could you also open a new issue with what's left to implement?

File: src/kontrol/kdist/cheatcodes.md (hunk 6 above)
Comment (lisandrasilva) [reply src/kontrol/kdist/cheatcodes.md:930]: I forgot to do that. Thanks!

File: src/kontrol/kdist/cheatcodes.md (hunk 7 above)
Comment (lisandrasilva) [reply src/kontrol/kdist/cheatcodes.md]: Done!

File: src/kontrol/kdist/cheatcodes.md [hunk 10]
```diff
@@ -907,6 +913,64 @@ Otherwise, throw an error for any other call to the Foundry contract.
       [owise]
 ```
 
+Mock calls
+----------
+
+#### `mockCall` - Mocks all calls to an address where if the call data either strictly or loosely
```
Comment (palinatolmach) [inline src/kontrol/kdist/cheatcodes.md:919]: IIUC, the convention is to have one line per sentence in Markdown files (@anvacaru please correct me if I'm wrong!).

File: src/kontrol/kdist/cheatcodes.md [hunk 11]
```diff
@@ -907,6 +913,62 @@ Otherwise, throw an error for any other call to the Foundry contract.
...
+
+#### `mockCall` - Mocks all calls to an address where if the call data either strictly or loosely matches data and returns retdata.
+
+```
+function mockCall(address where, bytes calldata data, bytes calldata retdata) external;
+```
+
+`foundry.call.mockCall` will match when the `mockCall` cheat code function is called.
+This rule then takes the `address` value from the function call data and etches a single byte into the account code, in case it is empty.
+The rule also takes the bytes `CALLDATA` and the bytes `RETURNDATA` from the function call data and forwards them, together with the address to be mocked, and forwards them to the `setMockCall` production.
```
Comment (palinatolmach) [inline src/kontrol/kdist/cheatcodes.md:927]: "Forwards them" is repeated twice:

```suggestion
The rule also takes the bytes `CALLDATA` and the bytes `RETURNDATA` from the function calldata and forwards them, together with the address to be mocked, to the `setMockCall` production.
```

File: src/kontrol/kdist/cheatcodes.md [hunk 12]
```diff
@@ -907,6 +913,62 @@ Otherwise, throw an error for any other call to the Foundry contract.
...
+----------
+
+#### `mockCall` - Mocks all calls to an address where if the call data either strictly or loosely matches data and returns retdata.
+
+```
+function mockCall(address where, bytes calldata data, bytes calldata retdata) external;
+```
+
+`foundry.call.mockCall` will match when the `mockCall` cheat code function is called.
+This rule then takes the `address` value from the function call data and etches a single byte into the account code, in case it is empty.
```
Comment (palinatolmach) [inline src/kontrol/kdist/cheatcodes.md:926]: It's a very minor thing, but I'd recommend to write "calldata" in one word to follow the Ethereum convention (here and below).

```suggestion
This rule then takes the `address` value from the function calldata and etches a single byte into the account code, in case it is empty.
```

File: src/kontrol/kdist/cheatcodes.md [hunk 13]
```diff
@@ -907,6 +913,62 @@ Otherwise, throw an error for any other call to the Foundry contract.
...
+         <k> #call_foundry SELECTOR ARGS
+          => #loadAccount #asWord(#range(ARGS, 0, 32))
+          ~> #etchAccountIfEmpty #asWord(#range(ARGS, 0, 32))
+          ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, #asWord(#range(ARGS, 32, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 32, 32)), 32))) #range(ARGS, #asWord(#range(ARGS, 64, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 64, 32)), 32)))
+         ...
+         </k> 
+      requires SELECTOR ==Int selector ( "mockCall(address,bytes,bytes)" )
+```
+
+We use the `#next[OP]` to identify OpCodes that represent function calls. If there is `<mockCall>` which `<mockAddress>`
```
Comment (palinatolmach) [inline src/kontrol/kdist/cheatcodes.md:944]: Shall this sentence be one-line as well? I think the wording here can also be a bit more streamlined, e.g., 
```suggestion
We use `#next[OP]` to identify OpCodes that represent function calls. If there is `<mockCall>`, for which `<mockAddress>` matches the `ACCTTO` and `<mockValues>` has a key `CALLDATA` that matches some prefix of the function calldata, then the `#execMockCall` will replace the function execution and update the output with the `RETURNDATA`.
```

File: src/kontrol/kdist/cheatcodes.md [hunk 14]
```diff
@@ -1316,6 +1378,58 @@ If the production is matched when no prank is active, it will be ignored.
...
+            => <mockCall>
+                  <mockAddress> MOCKADDRESS </mockAddress>
+                  <mockValues> .Map [ MOCKCALLDATA <- MOCKRETURN ] </mockValues>
+               </mockCall>
+           )
+           ...
+         </mockCalls>
+```
+
+- `#execMockCall` will update the output of the function call with `RETURNDATA` using `#setLocalMem` and in case the function
```
Comment (palinatolmach) [inline src/kontrol/kdist/cheatcodes.md:1422]: I'd split this into two sentences, e.g., something like
```suggestion
- `#execMockCall` will update the output of the function call with `RETURNDATA` using `#setLocalMem`.
In case the function did not end with `EVMC_SUCCESS` it will update the status code to `EVMC_SUCCESS`. 
```
Or you can just add a comma between two clauses.

Comment (lisandrasilva) [reply src/kontrol/kdist/cheatcodes.md]: Thanks for all the suggestions! I just addressed them.

File: src/kontrol/kdist/cheatcodes.md [hunk 15]
```diff
@@ -1316,6 +1376,57 @@ If the production is matched when no prank is active, it will be ignored.
         </whitelist>
 ```
 
+- `#etchAccountIfEmpty Account` - sets an Account code to a single byte '0u8' if the account is empty to circumvent the `extcodesize` check that Solidity might perform [source](https://github.com/foundry-rs/foundry/blob/b78289a0bc9df6e35624c632396e16f27d4ccb3f/crates/cheatcodes/src/evm/mock.rs#L54).
```
Comment (palinatolmach) [inline src/kontrol/kdist/cheatcodes.md:1379]: Sorry, I forgot to include it in the previous review. You can probably put the link to Foundry documentation in brackets as `(source)`, so it's more obvious that it's not part of the sentence :) 

```suggestion
- `#etchAccountIfEmpty Account` - sets an Account code to a single byte '0u8' if the account is empty to circumvent the `extcodesize` check that Solidity might perform ([source](https://github.com/foundry-rs/foundry/blob/b78289a0bc9df6e35624c632396e16f27d4ccb3f/crates/cheatcodes/src/evm/mock.rs#L54)).
```

**Review by palinatolmach — APPROVED** [review]

File: src/kontrol/kdist/cheatcodes.md [hunk 16]
```diff
@@ -77,6 +77,12 @@ module FOUNDRY-CHEAT-CODES
           <addressSet> .Set </addressSet>
           <storageSlotSet> .Set </storageSlotSet>
         </whitelist>
+        <mockCalls>
```
Comment (anvacaru) [inline src/kontrol/kdist/cheatcodes.md:80]: @lisandrasilva  One other thing I forgot to point out. Above this configuration there is a short description of each cell and its role. Could you update that as well?

File: LICENSE
Comment (anvacaru) [inline LICENSE:1]: This file doesn't belong in this PR as it has already been merged in master in #336.

File: src/tests/integration/test-data/foundry-prove-skip-legacy [hunk 17]
```diff
@@ -126,32 +143,40 @@ LoopsTest.test_sum_100()
 LoopsTest.test_sum_1000()
 LoopsTest.testSumToNBroken(uint256)
 LoopsTest.testSumToN(uint256)
-MockCallTest.testMockCall()
```
Comment (anvacaru) [inline src/tests/integration/test-data/foundry-prove-skip-legacy:129]: Can you time this proof on the legacy backend and skip it if it takes longer than 300s?

File: src/tests/integration/test-data/foundry-prove-all [hunk 18]
```diff
@@ -145,9 +145,14 @@ LoopsTest.test_sum_1000()
...
 MethodDisambiguateTest.test_method_call()
-MockCallTest.testMockCall()
+MockCallTest.testSelectorMockCall()
 MockCallTest.testMockCalls()
 MockCallTest.testMockCallValue()
+MockCallTestFoundry.testMockGetters()
+MockCallTestFoundry.testMockNested()
+MockCallTestFoundry.testMockSelector()
+MockCallTestFoundry.testMockCalldata()
+MockCallTestFoundry.testRevertMock()
```
Comment (anvacaru) [inline src/tests/integration/test-data/foundry-prove-all:155]: Similar to my previous comment. Could you time these on the legacy backend and add them to `foundry-prove-skip-legacy` if they take more than 300s?

File: src/kontrol/kdist/cheatcodes.md (hunk 16 above)
Comment (lisandrasilva) [reply src/kontrol/kdist/cheatcodes.md:80]: Sure :)

File: LICENSE
Comment (lisandrasilva) [reply LICENSE:1]: I think it's gone now I merged with master!

File: src/tests/integration/test-data/foundry-prove-all (hunk 18 above)
Comment (lisandrasilva) [reply src/tests/integration/test-data/foundry-prove-all:155]: Yes! I'll do that.

Comment (lisandrasilva) [reply src/tests/integration/test-data/foundry-prove-all:155]: I confirm that all of them take less than 300s. I added one more test that also passes and takes less than 300s. I added some more tests from `MockCallTest.sol` but added them to the skip lists since they are not passing yet.

File: src/kontrol/kdist/cheatcodes.md (hunk 16 above)
Comment (anvacaru) [reply src/kontrol/kdist/cheatcodes.md:83]: Thank you!

**Review by anvacaru — APPROVED** [review]
lgtm!
